core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -52,6 +57,177 @@ static TEE_Result ffa_get_dst(struct thread_smc_args *args,
 	return FFA_OK;
 }
 
+static int add_mem_region_to_sp(struct ffa_mem_access *mem_acc,
+				struct sp_mem *smem)
+{
+	return FFA_OK;
+}
+
+static void spmc_sp_handle_mem_share(struct thread_smc_args *args,
+				     struct ffa_rxtx *rxtx,
+				     struct sp_session *owner_sp)
+{
+	uint64_t global_handle = 0;
+	int res = FFA_OK;
+	uint32_t ret_w2 = 0;
+	uint32_t ret_w3 = 0;
+
+	cpu_spin_lock(&rxtx->spinlock);
+
+	/*clear top bits*/
+	args->a1 &= 0x0000ffff;
+
+	res = spmc_sp_add_share(rxtx, args->a1, &global_handle, owner_sp);
+	if (!res) {
+		reg_pair_from_64(global_handle, &ret_w3, &ret_w2);
+		args->a3 = ret_w3;
+		args->a2 = ret_w2;
+		args->a0 = FFA_SUCCESS_32;
+	} else {
+		ffa_set_error(args, res);
+	}
+
+	cpu_spin_unlock(&rxtx->spinlock);
+}
+
+static int spmc_sp_add_sp_region(struct sp_mem *smem,
+				 struct ffa_address_range *mem_reg,
+				 struct sp_session *owner_sp,
+				 uint8_t highest_permission)
+{
+	return FFA_OK;
+}
+
+static int spmc_sp_add_nw_region(struct sp_mem *smem,
+				 struct ffa_mem_region *mem_reg)
+{
+	return FFA_OK;
+}
+
+int spmc_sp_add_share(struct ffa_rxtx *rxtx,
+		      size_t blen, uint64_t *global_handle,
+		      struct sp_session *owner_sp)
+{
+	int res = FFA_INVALID_PARAMETERS;
+	unsigned int num_mem_accs = 0;
+	unsigned int i = 0;
+	struct ffa_mem_access *mem_acc = NULL;
+	size_t needed_size = 0;
+	size_t addr_range_offs = 0;
+	struct ffa_mem_region *mem_reg = NULL;
+	uint8_t highest_permission = 0;
+	struct sp_mem *smem = sp_mem_new();
+	struct ffa_mem_transaction *input_descr = rxtx->rx;
+
+	if (!smem)
+		return FFA_NO_MEMORY;
+
+	if (owner_sp) {
+		if (owner_sp->endpoint_id !=
+		    READ_ONCE(input_descr->sender_id)) {
+			res = FFA_DENIED;
+			goto cleanup;
+		}
+	} else {
+		if (sp_get_session(READ_ONCE(input_descr->sender_id))) {
+			res = FFA_DENIED;
+			goto cleanup;
+		}
+	}
+
+	num_mem_accs = READ_ONCE(input_descr->mem_access_count);
+	mem_acc = input_descr->mem_access_array;
+
+	if (!num_mem_accs) {
+		res = FFA_DENIED;
+		goto cleanup;
+	}
+
+	/* Store the ffa_mem_transaction */
+	smem->transaction.sender_id = READ_ONCE(input_descr->sender_id);
```

Comment:
What if `input_descr->sender_id` has changed since you checked it above?
What you've read with `READ_ONCE()` should be stored in a temporary variable if you need to use it more than once.

Reply:
Done